### ISSUE TITLE
ldap3.search response is not parsed correctly

### ISSUE THREAD

**Issue by colicab** [jupyterhub/ldapauthenticator#71]
So, I just started connecting JupyterHub to our LDAP server

I didn't seem to work right away (giving in credentials and JupyterHub would say 'incorrect username/password', so I dove into ldapauthenticator.py code to find the issue

First of all, I cloned from the master branch, as I read there are issues with the 1.1 release concerning the bind user/password.

my Jupyter config in respect to LDAP (in italics to mask our server specifics)

c.JupyterHub.authenticator_class = 'ldapauthenticator.LDAPAuthenticator'
c.LDAPAuthenticator.server_address = '_LDAP SERVER IP_'
c.LDAPAuthenticator.bind_dn_template = str('uid={username},dc=_OUR.DOMAIN_,dc=be')
c.LDAPAuthenticator.lookup_dn = True
c.LDAPAuthenticator.user_search_base = 'dc=_OUR.DOMAIN_,dc=be'
c.LDAPAuthenticator.user_attribute = 'uid'
c.LDAPAuthenticator.lookup_dn_search_user = str('cn=admin,dc=_OUR.DOMAIN_,dc=be')
c.LDAPAuthenticator.lookup_dn_search_password = '_admin password_'
c.LDAPAuthenticator.lookup_dn_user_dn_attribute = 'uid'

my username I tried to log in with was let's way _FRANK_
I used self.log.warn to print out the return of the function **resolve_username**
```
resolved_username = self.resolve_username(username)
self.log.warn(resolved_username)
```

[W 2018-05-22 22:53:49.869 JupyterHub ldapauthenticator:327] ['_FRANK_']

Notice it returns a list. 

Taking a step back I checked the entire response of the ldap3 search operation for username _FRANK_:
`[{'raw_dn': b'uid=FRANK,dc=OUR.DOMAIN,dc=be', 'dn': 'uid=FRANK,dc=OUR.DOMAIN,dc=be', 'raw_attributes': {'uid': [b'FRANK']}, 'attributes': {'uid': ['FRANK']}, 'type': 'searchResEntry'}]
`

So, the value  of the attribite uid is a list. At first I thought  I missed something in my configuration resulting in this output structure. However, this response is in accordance with ldap3 documentation: http://ldap3.readthedocs.io/searches.html#response
> attributes: a dictionary of returned attributes and their values. Values are list.

This results in a weird looking dn
```
userdn = dn.format(username=resolved_username)
self.log.warn(userdn)
```
[W 2018-05-22 22:45:21.661 JupyterHub ldapauthenticator:332] ['uid=['_FRANK_'],dc=_OUR.DOMAIN_,dc=be']

So, it seemed that would be an issue. To confirm this, I hardcoded the userdn variable right after this assignment and the login was successful.

`userdn =['uid=FRANK,dc=_OUR.DOMAIN_,dc=be']`

In conclusion: it looks like the uid attribute from the SEARCH operation response isn't properly parsed.
It should be enough to change line 244
`return conn.response[0]['attributes'][self.lookup_dn_user_dn_attribute]`
into
`return conn.response[0]['attributes'][self.lookup_dn_user_dn_attribute][0]`

to retrieve the element from the 1-element list

Cheers,
Philip

**Comment by dhirschfeld** [comment]
Hi @colicab,
To debug further could you run jupyterhub with the `--debug` switch and post the output of the logs. The current code does work for others but there may be something specific to your setup we're not taking into account.

**Comment by andreas-wolf** [comment]
Hi. I can confirm the problem.  The value of the attribute uid is a list '`raw_attributes': {'uid': [b'andreas']}`
Currently I can't add more information.
@colicab 's fix `return conn.response[0]['attributes'][self.lookup_dn_user_dn_attribute][0]` worked for me

**Comment by dhirschfeld** [comment]
Thanks for the confirmation @andreas-wolf. Unfortunately I don't think the fix is appropriate as-is since it will break for any implementation which returns a string rather than a list. It should be possible to support both cases though...

**Comment by metrofun** [comment]
Can confirm the same problem, wasted the whole day on it debugging tcp dumps
